plugins/org.jboss.tools.websockets.ui/src/org/jboss/tools/websockets/ui/internal/ca/MethodCompletionProposal.java

```diff
@@ -0,0 +1,269 @@
+import org.eclipse.jdt.core.dom.Modifier;
+import org.eclipse.jdt.core.dom.NodeFinder;
+import org.eclipse.jdt.core.dom.SimpleName;
+import org.eclipse.jdt.core.dom.rewrite.ASTRewrite;
+import org.eclipse.jdt.core.dom.rewrite.ITrackedNodePosition;
+import org.eclipse.jdt.core.dom.rewrite.ImportRewrite;
+import org.eclipse.jdt.core.dom.rewrite.ListRewrite;
+import org.eclipse.jdt.core.formatter.CodeFormatter;
+import org.eclipse.jdt.core.formatter.IndentManipulation;
+import org.eclipse.jdt.internal.corext.codemanipulation.CodeGenerationSettings;
+import org.eclipse.jdt.internal.corext.codemanipulation.StubUtility;
+import org.eclipse.jdt.internal.corext.javadoc.JavaDocLocations;
+import org.eclipse.jdt.internal.corext.util.CodeFormatterUtil;
+import org.eclipse.jdt.internal.ui.JavaPlugin;
+import org.eclipse.jdt.internal.ui.JavaPluginImages;
+import org.eclipse.jdt.internal.ui.javaeditor.ASTProvider;
+import org.eclipse.jdt.internal.ui.preferences.JavaPreferencesSettings;
+import org.eclipse.jdt.internal.ui.text.java.JavaTypeCompletionProposal;
+import org.eclipse.jdt.internal.ui.text.java.hover.JavadocBrowserInformationControlInput;
+import org.eclipse.jdt.ui.CodeGeneration;
+import org.eclipse.jdt.ui.SharedASTProvider;
+import org.eclipse.jface.internal.text.html.HTMLPrinter;
+import org.eclipse.jface.text.BadLocationException;
+import org.eclipse.jface.text.Document;
+import org.eclipse.jface.text.IDocument;
+import org.eclipse.jface.text.IRegion;
+import org.eclipse.jface.text.TextUtilities;
+import org.eclipse.jface.text.contentassist.ICompletionProposalExtension4;
+import org.eclipse.text.edits.MalformedTreeException;
+
+/**
+ * 
+ * @author Viacheslav Kabanovich
+ *
+ */
+public class MethodCompletionProposal extends JavaTypeCompletionProposal implements ICompletionProposalExtension4 {
+	private IJavaProject fJavaProject;
+	private WebsocketMethodInfo methodInfo;
+	private int position = 0;
+	private int prefixLength;
+
+	public MethodCompletionProposal(IJavaProject jproject, ICompilationUnit cu, 
+			WebsocketMethodInfo methodInfo,
+			int start, int length, int position, int prefixLength,
+			String completionProposal) {
+		super(completionProposal, cu, start, length, JavaPluginImages.get(JavaPluginImages.IMG_MISC_PUBLIC), methodInfo.displayName, 20000);
+		Assert.isNotNull(jproject);
+		Assert.isNotNull(methodInfo.methodName);
+		Assert.isNotNull(methodInfo.paramTypes);
+		Assert.isNotNull(cu);
+
+		this.methodInfo = methodInfo;
+		
+		this.position = position;
+		this.prefixLength = prefixLength;
+
+		fJavaProject = jproject;
+
+		setReplacementString(completionProposal);
+	}
+
+	@Override
+	public CharSequence getPrefixCompletionText(IDocument document, int completionOffset) {
+		return methodInfo.methodName;
+	}
+
+	private CompilationUnit getRecoveredAST(IDocument document, int offset, Document recoveredDocument) {
+		CompilationUnit ast= SharedASTProvider.getAST(fCompilationUnit, SharedASTProvider.WAIT_ACTIVE_ONLY, null);
```

Comment:
I used pattern from org.eclipse.jdt.internal.ui.text.java.OverrideCompletionProposal.
Our case is very similar to that one. They certainly had their reasons to do so, and it is a safe way, tested by everyone who ever used content assist to override a method.
Can we please keep that code?